File: gcj2013R3A.py
#!/usr/bin/env python
import sys

def solve():
    R = 37
    B,N = map(int,sys.stdin.readline().split())
    X = sorted(map(int,sys.stdin.readline().split())+[0]*(R-N))
    X.append(X[R-1]+B+1)
    p=0.0
    for i,j in [(i,j) for i in range(1,R+1) for j in range(i,R+1)]:
        h = min((B-(j-i)+sum(X[:j]))/j,X[j]-1)
        if h<=0 or h<X[i-1] or h+1<X[j-1]:
            continue
        u = h*i-sum(X[:i])
        d = (h+1)*(j-i)-sum(X[i:j])
        p=max(p,u*(R-1.0)/i-u-d)
    return "%.6f"%p

for x in range(int(sys.stdin.readline())):
    print("Case #"+str(x+1)+": "+str(solve()))

File: gcj2013R3A.hs
#!/usr/bin/env runghc
import Text.Printf
import Data.List
main = getContents >>= mapM_ putStrLn . zipWith caseno [1..] . solve . lines
caseno i r = "Case #"++show i++": "++r
solve (c:cs) = map (toString . answer) . take (read c) . analyze $ cs ::[String]
analyze (c1:c2:cs) = (b,n,map read . words $ c2) : analyze cs
  where [b,n] = map read . words $ c1
toString = printf "%.6f" ::Double->String
answer (b,n,x) = maximum (0.0:p) where
  r = 37
  y = replicate (r-n) 0 ++ sort x ++ [maximum x+b+1]
  p = [cost(i,j,h)| i<-[1..r], j<-[i..r],
       h<-[min ((b-(j-i)+sum(take j y))`div`j) (y!!j-1)],
       h>0, h>=y!!(i-1), h+1>=y!!(j-1)]
  cost (i,j,h) = u*(fromIntegral r-1.0)/fromIntegral i-u-d where
    u = fromIntegral (h*i-sum(take i y))
    d = fromIntegral ((h+1)*(j-i)-sum(drop i . take j $ y))
